### PR TITLE
[NFC][LoopVectorize] Introduce new getEstimatedRuntimeVF function

### PR DESCRIPTION
**Pull request by david-arm**
There are lots of places where we try to estimate the runtime
vectorisation factor based on the getVScaleForTuning TTI hook.
I've added a new getEstimatedRuntimeVF function and taught
several places in the vectoriser to use this new function.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -4976,11 +4973,7 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
       MaxInterleaveCount = ForceTargetMaxVectorInterleaveFactor;
   }
 
-  unsigned EstimatedVF = VF.getKnownMinValue();
-  if (VF.isScalable()) {
-    if (std::optional<unsigned> VScale = getVScaleForTuning(TheLoop, TTI))
-      EstimatedVF *= *VScale;
-  }
+  unsigned EstimatedVF = getEstimatedRuntimeVF(TheLoop, TTI, VF);
   assert(EstimatedVF >= 1 && "Estimated VF shouldn't be less than 1");
```
Comment (hazzlim) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:4977]: Does it make any sense to move this assert into `getEstimatedRuntimeVF` ?

Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:4977]: It's a good question. It looks like it was really just asserting that `getVScaleForTuning` returned a non-zero value. Maybe it is worth moving into `getEstimatedRuntimeVF`.

**Review by hazzlim — APPROVED** [review]
LGTM! :)

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -4300,6 +4300,17 @@ getVScaleForTuning(const Loop *L, const TargetTransformInfo &TTI) {
   return TTI.getVScaleForTuning();
 }
 
+static unsigned getEstimatedRuntimeVF(const Loop *L,
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:4303]: Would be gerat if you could add a comment here to document exactly what we mean by estimated runtime VF here

**Review by fhahn — COMMENTED** [review]
LGTM, thanks!

**Review by fhahn — APPROVED** [review]

**Comment by david-arm** [comment]
Rebase + added a TODO in isEpilogueVectorizationProfitable

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -4291,6 +4291,21 @@ getVScaleForTuning(const Loop *L, const TargetTransformInfo &TTI) {
   return TTI.getVScaleForTuning();
 }
 
+/// This functions attempts to return a value that represents the vectorization
```
Comment (hazzlim) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:4294]: Just a small nit:
```suggestion
/// This function attempts to return a value that represents the vectorization
```

**Comment by david-arm** [comment]
Rebase. Updated `LoopVectorizationPlanner::cost` to use new function.

**Pull request merged**